src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,25 +1,111 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class VersionConverter : JsonConverter<Version>
     {
+#if BUILDING_INBOX_LIBRARY
+        private const int MinimumVersionLength = 3; // 0.0
+
+        private const int MaximumVersionLength = 43; // 2147483647.2147483647.2147483647.2147483647
+
+        private const int MaximumEscapedVersionLength = JsonConstants.MaxExpansionFactorWhileEscaping * MaximumVersionLength;
+#endif
+
         public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw ThrowHelper.GetInvalidOperationException_ExpectedString(reader.TokenType);
+            }
+
+#if BUILDING_INBOX_LIBRARY
+            bool isEscaped = reader._stringHasEscaping;
+
+            int maxLength = isEscaped ? MaximumEscapedVersionLength : MaximumVersionLength;
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+            if (reader.HasValueSequence)
+            {
+                if (!JsonHelpers.IsInRangeInclusive(reader.ValueSequence.Length, MinimumVersionLength, maxLength))
+                {
+                    throw ThrowHelper.GetFormatException(DataType.Version);
+                }
+
+                Span<byte> stackSpan = stackalloc byte[isEscaped ? MaximumEscapedVersionLength : MaximumVersionLength];
```

Comment:
could `maxLength` be used here?

Reply:
I followed same approach as in TimeSpanConverter https://github.com/dotnet/runtime/blob/30226bff32769a2a5c26d6d45696dc78bc825c64/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs#L24-L38
As far as i know JIT can optimize stackallocing const amount of bytes, but i'm not sure if it can optimize with `maxLength`, since it can be only one of two const values, but by itself it isn't const.

Reply:
cc @GrabYourPitchforks do you reckon the JIT could be more optimal with the second pattern (N) [here](https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAICYCMBYAUKgZgAIMiBhIgbzyNpOIGMB7AOwmCIEsWOBZMAB4BRAQAcwbTqwBiYBsCYAnAOoALTgBsApkIgMwo7gHMiAXiIA2ANw06oxZwBuYYFqLM2Hbn0GcoAVygANS1FCCkWABktFiNgVTMiABYCG1w6InsnFzcPdi4eIn4BP0DdfVEtOBCwiOjY+MTikXFJGTkFFXVtcoNjIgAqIt8A4NDw1nq41TSM21pCEiSigAoAIyYmDS4IXsq4AEoqeYyiAGVWgB41gE9XAD4idjkAawuJROeGF7ANDSYGERbq4ANqcXZ6AxVIgAfmGJVGeyqNQmURi0yIIHhpTGtUm6PiAF1ZnQAL4nE6LVDLABy60223BSMOx3Sp1o3iIUEEU0a5iZkP2sOxiMFyPGdQJCSxxRxKMlDRmJwy7xY1zuWkeXzerU+wFev3+gOBWhB3IEvNUxJO5NwpKAA===)? `maxLength` could only be a `const` value.